### PR TITLE
added export check to splinepy/io/iges.py

### PR DESCRIPTION
**Pull request by mkofler96**
# Overview
iges export now checks parametric dimension of spline, if it is 3, then a value error is raised and a suggestion for a solution is presented

## Addressed issues
*  solved issue #272

## Showcase
calling the example from the issue now gives the following error:
```
ValueError                                Traceback (most recent call last)
[c:\Users\mkofler\miniconda3\envs\splinepy_dev\Lib\site-packages\splinepy\microstructure_fix.ipynb](file:///C:/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb) Cell 3 line 1
     [13](vscode-notebook-cell:/c%3A/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb#W0sZmlsZQ%3D%3D?line=12) microstructure.microtile = getattr(sp.microstructure.tiles, "NutTile3D")().create_tile()[0]
     [14](vscode-notebook-cell:/c%3A/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb#W0sZmlsZQ%3D%3D?line=13) microstructure.tiling = [8, 4, 1]
---> [16](vscode-notebook-cell:/c%3A/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb#W0sZmlsZQ%3D%3D?line=15) sp.io.iges.export("test.igs",microstructure.create().patches)
     [17](vscode-notebook-cell:/c%3A/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb#W0sZmlsZQ%3D%3D?line=16) ig = sp.io.iges.load("test.igs")
     [18](vscode-notebook-cell:/c%3A/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/microstructure_fix.ipynb#W0sZmlsZQ%3D%3D?line=17) ig.show(control_points=False)

File [c:\Users\mkofler\miniconda3\envs\splinepy_dev\Lib\site-packages\splinepy\io\iges.py:41](file:///C:/Users/mkofler/miniconda3/envs/splinepy_dev/Lib/site-packages/splinepy/io/iges.py:41), in export(fname, splines)
     39 for ispl, spline in enumerate(splines):
     40   if spline.para_dim == 3:
---> 41       raise ValueError(f"Volumetric {type(spline)} found at splines[{ispl}], which cannot be handled by the Iges file format \n"
     42                        "Use a boundary representation instead by calling e.g. spline.boundary_multipatch()")
     44 return splinepy_core.export_iges(fname, splines)

ValueError: Volumetric <class 'splinepy.bezier.Bezier'> found at splines[0], which cannot be handled by the Iges file format 
Use a boundary representation instead by calling e.g. spline.boundary_multipatch()
```
## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [ ] Added test(s), is this necessary?

### CODE REVIEW

File: splinepy/io/iges.py [hunk 1]
```diff
@@ -35,4 +35,11 @@ def export(fname, splines):
     --------
     None
     """
+    # check if any spline contains volume elements
+    for ispl, spline in enumerate(splines):
+        if spline.para_dim == 3:
```
Comment (j042) [inline splinepy/io/iges.py:40]:
```suggestion
        if spline.para_dim > 2 and spline.dim > 3:
```

Comment (j042) [reply splinepy/io/iges.py:40]: can you test two things with other CAD tools:
1. are spline.dim = 2 splines properly understood?
2. what happens with spline.dim = 1 splines?

File: splinepy/io/iges.py [hunk 2]
```diff
@@ -35,4 +35,11 @@ def export(fname, splines):
...
+            raise ValueError(
+                f"Volumetric {type(spline)} found at splines[{ispl}], which cannot be handled by the .iges file format \n"
```
Comment (j042) [inline splinepy/io/iges.py:42]: these lines are causing CIs to fail. Check [here](https://github.com/tataratat/splinepy/blob/main/CONTRIBUTING.md#automatic-formatting--style-check)

Comment (mkofler96) [reply splinepy/io/iges.py:42]: this should be fixed by the auto-import commit

Comment (j042) [reply splinepy/io/iges.py:42]: It's not

File: splinepy/io/iges.py [hunk 3]
```diff
@@ -35,4 +35,13 @@ def export(fname, splines):
...
     None
     """
+    # check if any spline contains volume elements
+    for ispl, spline in enumerate(splines):
+        if spline.para_dim > 2 or spline.dim > 3:
+            raise ValueError(
+                f"Volumetric {type(spline)} found at splines[{ispl}]"
+                "which cannot be handled by the .iges file format \n"
+                "Use a boundary representation instead by calling e.g."
+                "multipatch.boundary_multipatch()"
```
Comment (j042) [inline splinepy/io/iges.py:45]:
```suggestion
                "Use a boundary representation instead by calling e.g."
                "spline.extract.boundaries()"
```
This is applicable for both single and multipatch

File: splinepy/io/iges.py [hunk 4]
```diff
@@ -35,4 +35,13 @@ def export(fname, splines):
     --------
     None
     """
+    # check if any spline contains volume elements
+    for ispl, spline in enumerate(splines):
+        if spline.para_dim > 2 or spline.dim > 3:
```
Comment (j042) [inline splinepy/io/iges.py:40]: what about para_dim=2 and dim=1?

File: splinepy/io/iges.py [hunk 5]
```diff
@@ -35,4 +36,27 @@ def export(fname, splines):
...
+                f"Parametric dimension ({spline.para_dim}) is higher than the"
+                f" physical dimension ({spline.dim}) at splines[{ispl}]\n"
+                " Parametric dimension must be smaller or equal to the"
+                " physical dimension."
+            )
+        else:
+            valid_splines.append(spline)
+    if not valid_splines:
+        log.warning("No valid splines found in the given input")
+    return splinepy_core.export_iges(fname, valid_splines)
```
Comment (j042) [inline splinepy/io/iges.py:62]:
```suggestion
            valid_splines.append(spline)

    if len(valid_splines) == 0:
        raise ValueError("No valid splines found in the given input")

    return splinepy_core.export_iges(fname, valid_splines)
```
I'd prefer explicit length check. Wouldn't this make more sense to raise error here?

**Review by j042 — APPROVED** [review]
Thanks! Can you add a round-trip test? You can create a separate PR and we can merge both when the test is ready

**Comment by mkofler96** [comment]
> Thanks! Can you add a round-trip test? You can create a separate PR and we can merge both when the test is ready

Did you think of something like this? [tests/test_json.py
](https://github.com/tataratat/splinepy/blob/main/tests/test_json.py)

If yes: should we keep all import/export tests in a single file?
Then one could loop over every import/export method, something like
```python
        with tempfile.TemporaryDirectory() as tmpd:
            tmpf = c.to_tmpf(tmpd)
            
            export_formats = [(c.splinepy.io.json, {"bas64encoding": False}), (c.splinepy.io.iges, {}), (c.splinepy.io.irit, {}) ...]

            for (ef, ef_arguments) in export_formats: 
                ef.export(
                    tmpf, list_of_splines, **ef_arguments
                )
                list_of_splines_loaded = ef.load(tmpf)
                self.assertTrue(
                    all(
                        c.are_splines_equal(a, b)
                        for a, b in zip(list_of_splines, list_of_splines_loaded)
                    )
                )
```

What do you think?

**Comment by j042** [comment]
> > Thanks! Can you add a round-trip test? You can create a separate PR and we can merge both when the test is ready
> 
> Did you think of something like this? [tests/test_json.py ](https://github.com/tataratat/splinepy/blob/main/tests/test_json.py)
> 
> If yes: should we keep all import/export tests in a single file? Then one could loop over every import/export method, something like
> 
> ```python
>         with tempfile.TemporaryDirectory() as tmpd:
>             tmpf = c.to_tmpf(tmpd)
>             
>             export_formats = [(c.splinepy.io.json, {"bas64encoding": False}), (c.splinepy.io.iges, {}), (c.splinepy.io.irit, {}) ...]
> 
>             for (ef, ef_arguments) in export_formats: 
>                 ef.export(
>                     tmpf, list_of_splines, **ef_arguments
>                 )
>                 list_of_splines_loaded = ef.load(tmpf)
>                 self.assertTrue(
>                     all(
>                         c.are_splines_equal(a, b)
>                         for a, b in zip(list_of_splines, list_of_splines_loaded)
>                     )
>                 )
> ```
> 
> What do you think?

yes

**Comment by jzwar** [comment]
Is this still beeing worked on?

**Comment by mkofler96** [comment]
> Is this still beeing worked on?

yes, I'm still working on the round-trip tests, I just have not found the time for it yet

**Comment by j042** [comment]
so turns out BSplineLib will append zeros in case of dim < para_dim and always export 3D splines (maybe this is IGES rule?)

**Pull request merged**